Chjune android step1

engineer-choi opened the pull request:
Enum class부분 거의 똑같은 코드로 작성을 하게 되었네요..
그리고 기존의 검증 -> 계산 로직을 좀 더 세분화시켜 리팩터링 해봤습니다.

기존에는 전체 리스트를 돌며 숫자인지 연산자인지 검증을 완료하고 나서 계산을 수행했었으나,

수정 후에는 처음으로, 전체 사이즈만 검증해보도록 바꾸었습니다.
이후 for 구문을 돌며 숫자로 변환 가능한지, 연산 가능한지 검증과 계산을 동시에 하도록 적용해봤습니다.

engineer-choi committed:
[테스트입니다]
Step1 브랜치 파서 푸시 및 브랜치 생성 테스트입니다

engineer-choi committed:
[예외처리 테스트코드 완료]
사칙연산 기호가 아닌 &를 사용하였을 때 IllegalArgumentException 발생하도록 예외처리 완료

engineer-choi committed:
[사칙연산 테스트코드 완료]
+ - * / 사칙연산에 대한 테스트코드 작성 완료

engineer-choi committed:
[결과도출 완료]
+ - * / 사칙연산에 대한 결과값 출력 완료.
겁나 하드코딩이라 리팩터링 개많이 필요할듯

engineer-choi committed:
[1차 리팩터링 완료]
피드백 받은 예외처리 추가.
전역 메서드 -> 클래스별 역할 분배

engineer-choi committed:
[2차 리팩터링 완료]
연산자 검사에 대한 불필요한 낭비 줄이기 위해 index 2개씩 검사하는 로직 적용.
calculate()의 내부에 있던 분리, 검증 과정 분리.
피드백 반영하여 전체적 리팩터링

engineer-choi committed:
[2주차 리팩터링 완료]
기존 : 숫자와 연산자에 검사를 전부 하고 나서 계산을 수행하는 로직

engineer-choi committed:
[2주차 리팩터링 완료]
기존 : 숫자와 연산자에 검사를 전부 하고 나서 계산을 수행하는 로직
수정 : List의 Size체크 후 연산과 검증을 for문을 돌리는 동시에 실시하도록 함

engineer-choi committed:
Merge branch 'CHJUNE-android' into CHJUNE-android-step1

malibinYun commented on src/test/kotlin/OperatorTest.kt:
```diff
@@ -0,0 +1,30 @@
+        val right = 1
+        // then
+        assertThat(operator.calculateStrategy(left, right)).isEqualTo(result)
+    }
```
ParameterizedTest 👍👍

malibinYun commented on src/main/kotlin/SizeChecker.kt:
```diff
@@ -0,0 +1,7 @@
+    fun checkSize(splitList: List<String>) {
+        require(splitList.size > 2) { IllegalArgumentException("계산 식이 아닙니다.\n 입력 받은 값 : $splitList") }
+    }
+}
```
아예 문자열 분리해주는 애를 따로 만들어서 거기에 검증을 넣어주는건 어떨까?
지금은 결합도가 더 증가한거같아.

malibinYun commented on src/main/kotlin/NumChecker.kt:
```diff
@@ -0,0 +1,7 @@
+    fun checkNum(num: String): Int {
+        return num.toIntOrNull() ?: throw IllegalArgumentException("해당 문자는 숫자로 변환할 수 없습니다. \n 입력된 문자 : $num")
+    }
+}
```
얘도 결합도가 오히려 증가한 케이스인거같아.
calculator 안에 메서드로 존재해도 충분할거같아. 오히려 그게 응집도가 더 높고, 결합도를 줄이는 방법같아

malibinYun replied:
또,  검증을 했는데 해당 숫자가 반환되는것도 뭔가 어색해보여!

malibinYun commented on src/test/kotlin/StringCalculatorTest.kt:
```diff
@@ -1,10 +1,29 @@
 import org.junit.jupiter.api.Test
+import org.junit.jupiter.params.ParameterizedTest
 
+/**
+ * 2주차 리팩터링 사고과정
+ * 1 + 2 * 3 - 1 / 4
+ * 문자열로 받았다 -> 쪼개는 과정이 있어야 할 것이다.
+ *
+ * 쪼갠 문자열 하나씩 유효한 문자인지 검증하는 과정이 있어야 한다
+ *  -> 계산식이 맞는지 : 2 + 3 처럼 쪼갰을 때 배열의 최대 index가 2 이상 나와야 함
+ *  -> 홀수 index면 숫자인지
+ *  -> 짝수 index면 사칙연산자인지
+ *
+ * 계산하기
+ * 숫자 -> 연산자 -> 숫자 -> 연산자 -> 숫자 순서.
+ * 이전 계산 결과 데이터에
+ *      해당 연산자에 맞는 계산으로 다음 숫자와 연산된 값을 결과에 대입.
+ *
+ * 문자열 입력 -> 쪼갬 -> 검증 -> 계산 -> 출력 5단계 과정
+ *
+ */
```
요구사항 분석 👍👍👍

malibinYun commented on src/main/kotlin/StringCalculator.kt:
```diff
@@ -1,18 +1,17 @@
 class StringCalculator {
     fun calculate(splitList: List<String>): Int {
-        var resultNum = splitList[0].toInt()
+        val numChecker = NumChecker()
```
NumChecker() 는 calculate 할 때 마다 새로 생성되는데, 만약 NumChecker()생성 비용이 매우 크다면 어떤 문제가 발생할까?

engineer-choi replied:
OOM이 발생하려남..?

malibinYun replied:
그럴수도있고!
생성해서 1회용 객체가 아닌이상 딱 한 번만 만들어두고 사용하는게 더 효율적일거같아!

malibinYun commented on src/test/kotlin/OperatorTest.kt:
```diff
@@ -0,0 +1,30 @@
+    @ParameterizedTest
+    @CsvSource("PLUS,4", "MINUS,2", "MULTIPLY,3", "DIVIDE,3")
+    fun plusTest(operator: Operator, result: Int) {
+        // given
+
+        // when
+        val left = 3
+        val right = 1
+        // then
+        assertThat(operator.calculateStrategy(left, right)).isEqualTo(result)
```
```suggestion
        // given
        val left = 3
        val right = 1
        
        // then
        assertThat(operator.calculateStrategy(left, right)).isEqualTo(result)
```

malibinYun reviewed (approved):
몬가 많이 개선이 되었군요!
요구사항 분석 아주 좋았어 굳굳 

> 기존에는 전체 리스트를 돌며 숫자인지 연산자인지 검증을 완료하고 나서 계산을 수행했었으나,
수정 후에는 처음으로, 전체 사이즈만 검증해보도록 바꾸었습니다.
이후 for 구문을 돌며 숫자로 변환 가능한지, 연산 가능한지 검증과 계산을 동시에 하도록 적용해봤습니다.

이것도 좋은 접근방법인거같아!!
더 리팩터링 하려면 해도되구 그만하려면 머지해달라구 알려줘. 우선은 풀리퀘 열어둘게

engineer-choi committed:
[2주차 리팩토링 수정]
결합도 낮추고 응집도 높이도록 수정
문자열 분리와 size검증에 대한 로직은 함께 진행
계산 과정에서 사칙연산과 숫자 검증을 함께 진행하도록 수정
